src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs

```diff
@@ -47,15 +47,17 @@ private ArrayList InnerList
             {
                 if (_innerList == null)
                 {
-                    _innerList = new ArrayList();
+                    var eagerList = new ArrayList();
                     var enumerator = new ResultsEnumerator(
                         this,
                         _rootEntry.GetUsername(),
                         _rootEntry.GetPassword(),
                         _rootEntry.AuthenticationType);
 
                     while (enumerator.MoveNext())
-                        _innerList.Add(enumerator.Current);
+                        eagerList.Add(enumerator.Current);
+
+                    _innerList = eagerList;
```

Comment:
The change to build the inner list in it's entirety _and then_ set the member is to ensure that if an exception in thrown while populating this inner list we do NOT memoize the incomplete list because then we would (if enumerated again) not throw again and would instead silently only yield the entries we previously gathered. Of course anyone attempting _re_enumeration of a list that yielded an error the first time is probably wrong, but why add to the problem?